File: evaluations/powersimulations/tests/expressiveness/test_a4_ac_feasibility.jl
#=
Test A-4: AC Feasibility Check on DC OPF Dispatch

Dimension: expressiveness
Network: TINY (IEEE 39-bus)
Pass condition: Achievable within same model context (no export/reimport).
  Voltage and thermal violations identifiable.
Tool: PowerSimulations.jl v0.30.2
=#

using PowerSystems
using PowerSimulations
using PowerFlows
using HiGHS
using Ipopt
using JuMP
using JSON
using Logging
using DataFrames
using CSV
using Dates
using TimeSeries: TimeArray

# Suppress verbose logging
global_logger(ConsoleLogger(stderr, Logging.Error))

const PSI = PowerSimulations

function peak_rss_mb()
    for line in eachline("/proc/self/status")
        if startswith(line, "VmHWM:")
            return parse(Float64, split(line)[2]) / 1024  # kB to MB
        end
    end
    return nothing
end

# Cost mapping from README
const COST_MAP = Dict("hydro" => 5.0, "nuclear" => 10.0, "coal_large" => 25.0, "gas_CC" => 40.0)

function setup_system(network_file, timeseries_dir)
    sys = System(network_file)

    # Apply differentiated costs (same as A-3)
    params = CSV.read(joinpath(timeseries_dir, "gen_temporal_params.csv"), DataFrame)
    for row in eachrow(params)
        c1 = get(COST_MAP, row.tech_class_key, 30.0)
        c2 = c1 * 0.001
        for gen in get_components(ThermalStandard, sys)
            if get_number(get_bus(gen)) == row.bus_id
                set_operation_cost!(
                    gen,
                    ThermalGenerationCost(CostCurve(QuadraticCurve(c2, c1, 0.0)), 0.0, 0.0, 0.0),
                )
                break
            end
        end
    end

    # Apply 70% branch derating (same as A-3)
    for line in get_components(Line, sys)
        set_rating!(line, get_rating(line) * 0.7)
    end
    for xfmr in get_components(Transformer2W, sys)
        set_rating!(xfmr, get_rating(xfmr) * 0.7)
    end
    for xfmr in get_components(TapTransformer, sys)
        set_rating!(xfmr, get_rating(xfmr) * 0.7)
    end

    # Add time series (required by PowerSimulations)
    timestamps = [DateTime("2024-01-01"), DateTime("2024-01-01") + Hour(1)]
    for load in get_components(PowerLoad, sys)
        add_time_series!(
            sys, load, SingleTimeSeries("max_active_power", TimeArray(timestamps, [1.0, 1.0]))
        )
    end
    transform_single_time_series!(sys, Hour(1), Hour(1))

    return sys
end

function run_dcopf(sys, solver)
    template = ProblemTemplate(NetworkModel(DCPPowerModel; duals=[NodalBalanceActiveConstraint]))
    set_device_model!(template, ThermalStandard, ThermalDispatchNoMin)
    set_device_model!(template, PowerLoad, StaticPowerLoad)
    set_device_model!(template, Line, StaticBranch)
    set_device_model!(template, Transformer2W, StaticBranch)
    set_device_model!(template, TapTransformer, StaticBranch)

    model = DecisionModel(template, sys; optimizer=solver)
    build!(model; output_dir=mktempdir())
    solve!(model)
    return model
end

function run(
    network_file::String="/workspace/data/networks/case39.m";
    timeseries_dir::Union{String,Nothing}="/workspace/data/timeseries/case39",
)
    results = Dict(
        "status" => "fail",
        "wall_clock_seconds" => 0.0,
        "details" => Dict(),
        "errors" => String[],
        "workarounds" => String[],
    )

    try
        if timeseries_dir === nothing
            push!(results["errors"], "timeseries_dir required for A-4")
            return results
        end

        highs_solver = optimizer_with_attributes(
            HiGHS.Optimizer,
            "time_limit" => 300.0,
            "presolve" => "on",
            "threads" => 1,
            "output_flag" => false,
        )

        base_power_val = nothing

        # Warm-up
        sys_w = setup_system(network_file, timeseries_dir)
        run_dcopf(sys_w, highs_solver)

        # Timed run
        sys = setup_system(network_file, timeseries_dir)
        base_power_val = get_base_power(sys)

        t0 = time()

        # Step 1: Solve DCOPF
        model = run_dcopf(sys, highs_solver)
        res = OptimizationProblemResults(model)

        # Extract DCOPF dispatch
        dispatch_df = read_variable(res, "ActivePowerVariable__ThermalStandard")
        dcopf_dispatch = Dict{String,Float64}()
        for col in names(dispatch_df)
            col == "DateTime" && continue
            dcopf_dispatch[col] = dispatch_df[1, col]
        end

        results["details"]["dcopf_dispatch_mw"] = Dict(
            k => round(v; digits=2) for (k, v) in dcopf_dispatch
        )
        results["details"]["base_power_mva"] = base_power_val

        # Step 2: Fix generator active power to DCOPF dispatch values (in-place on System)
        gen_updates = Dict{String,Any}[]
        for gen in get_components(ThermalStandard, sys)
            gname = get_name(gen)
            if haskey(dcopf_dispatch, gname)
                dispatch_mw = dcopf_dispatch[gname]
                # PowerSystems stores active_power in per-unit (MW / base_power)
                dispatch_pu = dispatch_mw / base_power_val
                set_active_power!(gen, dispatch_pu)
                push!(
                    gen_updates,
                    Dict(
                        "name" => gname,
                        "dispatch_mw" => round(dispatch_mw; digits=2),
                        "dispatch_pu" => round(dispatch_pu; digits=4),
                    ),
                )
            end
        end
        results["details"]["gen_updates"] = gen_updates

        # Step 3: Run ACPF using PowerFlows.jl (same System object — no export/reimport)
        pf_result = solve_powerflow(ACPowerFlow(), sys)

        elapsed = time() - t0
        results["wall_clock_seconds"] = elapsed
        results["details"]["peak_memory_mb"] = peak_rss_mb()

        # Step 4: Analyze results
        if pf_result === nothing
            push!(results["errors"], "AC power flow did not converge")
            results["details"]["acpf_converged"] = false
        else
            results["details"]["acpf_converged"] = true

            # Extract bus voltages and check violations
            bus_df = pf_result["bus_results"]
            voltage_violations = Dict{String,Any}[]
            for row in eachrow(bus_df)
                vm = row[:Vm]
                if vm < 0.95 || vm > 1.05
                    push!(
                        voltage_violations,
                        Dict(
                            "bus" => row[:bus_number],
                            "voltage_pu" => round(vm; digits=4),
                            "violation" => vm < 0.95 ? "under" : "over",
                            "deviation_pu" =>
                                round(vm < 0.95 ? (0.95 - vm) : (vm - 1.05); digits=4),
                        ),
                    )
                end
            end
            results["details"]["voltage_violations"] = voltage_violations
            results["details"]["num_voltage_violations"] = length(voltage_violations)

            # Voltage summary
            vm_vals = bus_df[:, :Vm]
            results["details"]["voltage_summary"] = Dict(
                "min_pu" => round(minimum(vm_vals); digits=4),
                "max_pu" => round(maximum(vm_vals); digits=4),
                "mean_pu" => round(sum(vm_vals) / length(vm_vals); digits=4),
                "n_buses" => length(vm_vals),
            )

            # Extract branch flows and check thermal violations
            flow_df = pf_result["flow_results"]
            thermal_violations = Dict{String,Any}[]
            for row in eachrow(flow_df)
                # Flow magnitude
                pf = row[:P_from_to]
                qf = row[:Q_from_to]
                s_flow = sqrt(pf^2 + qf^2)

                # Get branch rating
                bname = row[:line_name]
                rating_mw = nothing
                for line in get_components(Line, sys)
                    if get_name(line) == bname
                        rating_mw = get_rating(line) * base_power_val
                        break
                    end
                end
                if rating_mw === nothing
                    for xfmr in get_components(Transformer2W, sys)
                        if get_name(xfmr) == bname
                            rating_mw = get_rating(xfmr) * base_power_val
                            break
                        end
                    end
                end
                if rating_mw === nothing
                    for xfmr in get_components(TapTransformer, sys)
                        if get_name(xfmr) == bname
                            rating_mw = get_rating(xfmr) * base_power_val
                            break
                        end
                    end
                end

                if rating_mw !== nothing && rating_mw > 0 && s_flow > rating_mw
                    loading_pct = s_flow / rating_mw * 100.0
                    push!(
                        thermal_violations,
                        Dict(
                            "branch" => bname,
                            "flow_mva" => round(s_flow; digits=2),
                            "rating_mva" => round(rating_mw; digits=2),
                            "loading_pct" => round(loading_pct; digits=1),
                            "overload_mva" => round(s_flow - rating_mw; digits=2),
                        ),
                    )
                end
            end
            results["details"]["thermal_violations"] = thermal_violations
            results["details"]["num_thermal_violations"] = length(thermal_violations)

            # Reactive power summary
            gen_q_violations = Dict{String,Any}[]
            for gen in get_components(ThermalStandard, sys)
                gname = get_name(gen)
                q_pu = get_reactive_power(gen)
                q_mw = q_pu * base_power_val
                q_lims = get_reactive_power_limits(gen)
                q_min_mw = q_lims.min * base_power_val
                q_max_mw = q_lims.max * base_power_val
                if q_mw < q_min_mw - 0.01 || q_mw > q_max_mw + 0.01
                    push!(
                        gen_q_violations,
                        Dict(
                            "gen" => gname,
                            "q_mvar" => round(q_mw; digits=2),
                            "q_min_mvar" => round(q_min_mw; digits=2),
                            "q_max_mvar" => round(q_max_mw; digits=2),
                        ),
                    )
                end
            end
            results["details"]["reactive_power_violations"] = gen_q_violations
            results["details"]["num_reactive_violations"] = length(gen_q_violations)
        end

        # Pass condition: achievable within same model context, violations identifiable
        acpf_ok = get(results["details"], "acpf_converged", false)
        can_identify_voltage = haskey(results["details"], "voltage_violations")
        can_identify_thermal = haskey(results["details"], "thermal_violations")

        results["details"]["pass_checks"] = Dict(
            "same_model_context" => true,  # Used same System object
            "acpf_converged" => acpf_ok,
            "voltage_violations_identifiable" => can_identify_voltage,
            "thermal_violations_identifiable" => can_identify_thermal,
        )

        push!(
            results["workarounds"],
            "PowerSimulations DCOPF and PowerFlows ACPF share the same PowerSystems.System " *
            "object. DCOPF dispatch values are set in-place on generators via set_active_power!, " *
            "then ACPF runs on the same System. No export/reimport needed.",
        )

        if acpf_ok && can_identify_voltage && can_identify_thermal
            results["status"] = "pass"
        elseif !acpf_ok
            # ACPF failed to converge — still may be a qualified pass if we can show the workflow
            results["status"] = "qualified_pass"
            push!(
                results["workarounds"],
                "ACPF did not converge with DC dispatch fixed. This may indicate " *
                "reactive power infeasibility with fixed P dispatch. The workflow " *
                "itself (fix P from DCOPF, run ACPF) is supported.",
            )
        end

    catch e
        push!(results["errors"], string(typeof(e), ": ", sprint(showerror, e)))
        results["details"]["traceback"] = sprint(showerror, e, catch_backtrace())
    end

    return results
end

if abspath(PROGRAM_FILE) == @__FILE__
    result = run()
    println(JSON.json(result, 2))
end

File: evaluations/gridcal/tests/expressiveness/test_a4_ac_feasibility.py
"""
Test A-4: Take DC OPF dispatch from A-3, run full ACPF on that dispatch.

Dimension: expressiveness
Network: TINY (IEEE 39-bus New England)
Pass condition: Achievable within the same model context (no export to file and reimport).
    Voltage violations and thermal limit violations identifiable from results.
Tool: gridcal (VeraGridEngine) 5.6.28

Solver config specifies Ipopt, but GridCal has NO Ipopt integration for ACPF.
GridCal uses its own Newton-Raphson solver (SolverType.NR). This is documented as a finding.
"""

from __future__ import annotations

import csv
import json
import sys
import time
import traceback
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent.parent / "shared"))

from matpower_loader import load_gridcal

COST_MAP = {
    "hydro": {"c1": 5.0, "c2": 0.005},
    "nuclear": {"c1": 10.0, "c2": 0.010},
    "coal_large": {"c1": 25.0, "c2": 0.025},
    "gas_CC": {"c1": 40.0, "c2": 0.040},
}

BRANCH_DERATING = 0.70


def run(
    network_file: str = "data/networks/case39.m",
    timeseries_dir: str | None = "data/timeseries/case39",
) -> dict:
    """Execute A-4 AC feasibility test and return structured results."""
    results = {
        "status": "fail",
        "wall_clock_seconds": 0.0,
        "details": {},
        "errors": [],
        "workarounds": [],
    }

    start = time.perf_counter()
    try:
        import VeraGridEngine as vge
        from VeraGridEngine.enumerations import MIPSolvers, SolverType

        # 1. Load network
        grid = load_gridcal(network_file)
        generators = grid.get_generators()
        branches = grid.get_branches()
        buses = grid.get_buses()
        n_buses = grid.get_bus_number()
        n_gens = len(generators)
        results["details"]["bus_count"] = n_buses
        results["details"]["gen_count"] = n_gens
        results["details"]["branch_count"] = len(branches)

        # 2. Apply differentiated costs from Modified Tiny data
        ts_dir = Path(timeseries_dir) if timeseries_dir else None
        if ts_dir and (ts_dir / "gen_temporal_params.csv").exists():
            gen_params = {}
            with open(ts_dir / "gen_temporal_params.csv") as f:
                reader = csv.DictReader(f)
                for row in reader:
                    gen_params[int(row["gen_index"])] = row

            for idx, gen in enumerate(generators):
                if idx in gen_params:
                    tech_key = gen_params[idx]["tech_class_key"]
                    if tech_key in COST_MAP:
                        gen.Cost = COST_MAP[tech_key]["c1"]
                        gen.Cost2 = COST_MAP[tech_key]["c2"]
                        gen.Cost0 = 0.0
        else:
            results["errors"].append("gen_temporal_params.csv not found")

        # 3. Apply 70% branch derating (same as A-3)
        for branch in branches:
            if hasattr(branch, "rate") and branch.rate > 0:
                branch.rate = branch.rate * BRANCH_DERATING

        # 4. Solve DC OPF (same as A-3)
        opf_opts = vge.OptimalPowerFlowOptions(
            solver=SolverType.LINEAR_OPF,
            mip_solver=MIPSolvers.HIGHS,
        )
        opf_results = vge.linear_opf(grid, opf_opts)

        if not opf_results.converged:
            results["errors"].append("DC OPF (A-3 prerequisite) did not converge")
            return results

        dispatch = opf_results.generator_power
        results["details"]["dcopf_converged"] = True
        results["details"]["dcopf_dispatch_mw"] = {
            generators[i].name or f"gen_{i}": float(dispatch[i]) for i in range(n_gens)
        }
        results["details"]["dcopf_total_gen_mw"] = float(np.sum(dispatch))

        # 5. Fix generator dispatch to DC OPF values — WITHIN SAME MODEL CONTEXT
        #    No export/reimport. Just set gen.P and run ACPF on the same grid object.
        for i, gen in enumerate(generators):
            gen.P = float(dispatch[i])

        results["details"]["same_model_context"] = True
        results["details"]["solver_note"] = (
            "GridCal has no Ipopt integration. Using native NR solver (SolverType.NR)."
        )

        # 6. Run ACPF with flat start
        pf_opts = vge.PowerFlowOptions(
            solver_type=SolverType.NR,
            tolerance=1e-6,
            max_iter=100,
        )
        pf_results = vge.power_flow(grid, options=pf_opts)

        converged = bool(pf_results.converged)
        iterations = int(pf_results.iterations) if hasattr(pf_results, "iterations") else None
        residual = float(pf_results.error) if hasattr(pf_results, "error") else None
        results["details"]["acpf_converged"] = converged
        results["details"]["acpf_iterations"] = iterations
        results["details"]["acpf_residual"] = residual

        if not converged:
            # DC warm start fallback per convergence protocol
            dc_opts = vge.PowerFlowOptions(solver_type=SolverType.Linear)
            dc_results = vge.power_flow(grid, options=dc_opts)
            if dc_results.converged:
                for i, bus in enumerate(buses):
                    bus.Va0 = float(np.angle(dc_results.voltage[i]))
                pf_opts_warm = vge.PowerFlowOptions(
                    solver_type=SolverType.NR,
                    tolerance=1e-6,
                    max_iter=200,
                    use_stored_guess=True,
                )
                pf_results = vge.power_flow(grid, options=pf_opts_warm)
                converged = bool(pf_results.converged)
                results["details"]["dc_warmstart_needed"] = True
                results["details"]["dc_warmstart_converged"] = converged

        if not converged:
            results["errors"].append("ACPF did not converge even with DC warm start")
            return results

        # 7. Extract voltage results
        voltage = pf_results.voltage
        v_mag = np.abs(voltage)
        np.degrees(np.angle(voltage))

        results["details"]["vm_min"] = float(np.min(v_mag))
        results["details"]["vm_max"] = float(np.max(v_mag))

        # 8. Identify voltage violations (outside [0.95, 1.05] pu)
        bus_names = [b.name for b in buses]
        voltage_violations = []
        for i in range(n_buses):
            if v_mag[i] < 0.95 or v_mag[i] > 1.05:
                voltage_violations.append(
                    {
                        "bus": bus_names[i],
                        "vm_pu": float(v_mag[i]),
                        "violation": "under" if v_mag[i] < 0.95 else "over",
                    }
                )
        results["details"]["voltage_violations"] = voltage_violations
        results["details"]["voltage_violation_count"] = len(voltage_violations)

        # 9. Identify thermal violations (loading > 100% of derated rating)
        loading = np.abs(pf_results.loading)
        branch_names = [b.name for b in branches]
        thermal_violations = []
        for i in range(len(loading)):
            if loading[i] > 1.0:
                thermal_violations.append(
                    {
                        "branch": branch_names[i] if i < len(branch_names) else f"branch_{i}",
                        "loading_pct": float(loading[i] * 100),
                    }
                )
        results["details"]["thermal_violations"] = thermal_violations
        results["details"]["thermal_violation_count"] = len(thermal_violations)

        # 10. Reactive power information
        losses = pf_results.losses
        total_losses_mw = float(np.sum(np.real(losses)))
        results["details"]["total_losses_mw"] = total_losses_mw

        # 11. Check pass conditions
        pass_checks = {
            "dcopf_converged": True,
            "acpf_converged": converged,
            "same_model_context": True,  # no export/reimport needed
            "violations_identifiable": len(voltage_violations) > 0 or len(thermal_violations) > 0,
        }
        results["details"]["pass_checks"] = pass_checks

        if all(pass_checks.values()):
            results["status"] = "pass"
        else:
            failing = [k for k, v in pass_checks.items() if not v]
            results["errors"].append(f"Failed checks: {failing}")

    except Exception as e:
        results["errors"].append(f"{type(e).__name__}: {e}")
        results["details"]["traceback"] = traceback.format_exc()
    finally:
        results["wall_clock_seconds"] = time.perf_counter() - start

    return results


if __name__ == "__main__":
    result = run()
    print(json.dumps(result, indent=2, default=str))
